core/services/pipeline/orm.go

```diff
@@ -192,7 +192,7 @@ func (o *orm) UpdateTaskRunResult(db *sql.DB, taskID uuid.UUID, result interface
 // If saveSuccessfulTaskRuns = false, we only save errored runs.
 // That way if the job is run frequently (such as OCR) we avoid saving a large number of successful task runs
 // which do not provide much value.
-func (o *orm) InsertFinishedRun(db *gorm.DB, run Run, trrs []TaskRunResult, saveSuccessfulTaskRuns bool) (runID int64, err error) {
+func (o *orm) InsertFinishedRun(ctx context.Context, run Run, trrs []TaskRunResult, saveSuccessfulTaskRuns bool) (runID int64, err error) {
```

Comment:
I'd rather see changes to pipeline/orm.go get reverted because they will conflict #4775 where this method is rewritten to use sqlx.

Reply:
@archseer in this case, I need to revert almost all of that stuff I did because all of these things are related to each other. Anyway, we really need to avoid passing any specific DB-related objects to ORM. ONLY ORM layer should know its implementation details and which kind of database it uses.

Reply:
Right, but all of this code predates `TxFromContext` -- we're hoping to gradually drop `gorm` altogether.

(I was also told to avoid `context.WithValue`).

In particular I'm not sure I agree with this:

> It is better to use context with some DB-related value either than use the PostgreSQL DB connection in the service layer.

Hiding the DB in the context does make the API nicer, but now bugs stop being compile time and instead fail at runtime. For instance, incorrectly passing a `gorm.DB` value context into a function that expects `sqlx.Txx` will now silently fail and use the ORM's `sqlx.DB`, running outside of the intended transaction.

I'd rather see the ORM stops being a stateful struct altogether and turns into a set of functions that expect me to provide either a `sqlx.DB` or a `sqlx.Txx` (I've abstracted a `postgres.Queryer` interface that covers both).

Reply:
Agree with hidden DB object passing. But ORM interface should be independent of any specific databases. Today we use postgres, but what if tomorrow we want to use ES or another DB? How much impact will be? Our ORM interfaces are strictly dependent on s specific DB which is not good. The goal of an interface is to describe the behavior of a specific layer and it should not contain anything related to a specific implementation.